ksmt-core/src/main/kotlin/org/ksmt/expr/Array.kt

```diff
@@ -31,6 +35,286 @@ sealed class KArrayStoreBase<A : KArraySortBase<R>, R : KSort>(
             addAll(this@KArrayStoreBase.indices)
             add(value)
         }.uncheckedCast()
+
+    /**
+     * Array store indices cache.
+     *
+     * Each node in array store chain contains a cache for only one index.
+     * Cached index is specified by the [cacheKind] property.
+     * Cached indices are rotated.
+     * For example, in a case of 3 dimensional array nodes [cacheKind] will be 0, 1, 2, 0, 1, 2 ...
+     *
+     * If array store was not analyzed yet [cacheKind] is [NOT_INITIALIZED_CACHE_KIND].
+     * */
+    private var cacheKind: Int = NOT_INITIALIZED_CACHE_KIND
+
+    /**
+     * Table to perform fast index lookups.
+     * Can be `null` in the following cases:
+     * 1. Store was not analyzed.
+     * 2. Store index is uninterpreted.
+     * 3. Stores chain is small (see [LINEAR_LOOKUP_THRESHOLD]) and linear lookups are used.
+     * */
+    private var cacheLookupTable: PersistentMap<KExpr<*>, KArrayStoreBase<A, *>>? = null
```

Comment:
Currently, you can disable cache via overriding `mkArrayStoreNoSimplify` in `KContext`.

Reply:
That's right. We decided to implement two ways of decoding `KModel` in the USVM:

- Decode `KModel` to our entities eagerly, so it can be garbage collected (here we don't need KSMT array cache)
- Detach `KModel` and evaluate values from it explicitly (here we do need KSMT array cache)

Reply:
If you want to manually control array store caching you can do the following:
1. Disable cache by default via overriding `mkArrayStoreNoSimplify`.
2. Manually invoke `.analyzeStore()` on a store expressions when you need some caching.